README
!-----------------------------------------------------------------
! This file is part of
!
!  Trajman: A MD Trajectory Analysis Tool
!
!  Copyright (C) 2010, Jon Kapla, jon.kapla@mmk.su.se
!  Department of Material and Environmental Chemistry,
!  Stockholm University
!
! Trajman is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.
!
! Trajman is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License
! along with Trajman.  If not, see <http://www.gnu.org/licenses/>.
!-----------------------------------------------------------------

TRAJMAN: A MD TRAJECTORY ANALYSIS TOOL

Trajman is a linux based molecular dynamics trajectory analysis
tool written in fortran. It is mainly aimed towards analysis of
simulations of model lipid membranes, but can easily be adopted
for other use as well.

Get Trajman source code:

Trajman source code can be downloaded with bazaar. Get a recent
version of bazaar from http://bazaar.canocical.com and issue the 
following command:

> bzr branch http://people.su.se/~jkapl/trajman trajman

Since the program is in constant development it might be wise
to check the commit log for suitable revisions and changes.

> bzr log -n0

Dependencies:

Trajman needs the Visual Molecular Dynamics molfile_plugin to
be functional. There is a version of the molfile_plugin 
in the trajman build dir. You have also the possibility to 
download and build the latest version from:

http://www.ks.uiuc.edu/Research/vmd/

Make sure you also compile the fortran c wrapper located in 
the molfile_plugin/f77 folder.

For dependencies of the vmd plugins, please see the vmd homepage.

BUILD EXAMPLE:

Compile vmd plugins (Please read the license file in the
vmd_plugins1.9 dir):

> cd vmd_plugins1.9
> make LINUXAMD64 TCLINC=-I/usr/include/tcl TCLLIB=-F/usr/lib
> cd molfile_plugin/f77
> make ARCH=LINUXAMD64 FC=gfortran TCLLIB=-L/usr/lib NETCDFLIB=-L/usr/lib TCLLDFLAGS=-ltcl8.5 f77_molfile.o

When you have VMD plugins compiled. Compile trajman by telling
make where your VMD plugins are situated, and what archflag you
used. Note that you need gfortran. Here's a 64-bit example:

make VMD_ARCH=LINUXAMD64 VMD_PLUGINS=vmd_plugins1.9/
(or set the environment variables in the Makefile directly and use 'make')

Run trajman -h for more information about usage.


changelog.txt
------------------------------------------------------------
revno: 22 [merge]
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Wed 2011-10-05 13:59:55 +0200
message:
  Merged with trajsplit. Changes from Baltzar. See bzr log -n0.
    ------------------------------------------------------------
    revno: 18.1.3 [merge]
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: trajsplit
    timestamp: Wed 2011-10-05 13:56:52 +0200
    message:
      Merged Baltzars changes.
       +HISTORY trajectory support (use HISTORY file as init).
       +Changes in the RDF routine.
       +Updated Plancks constant.
       
        ------------------------------------------------------------
        revno: 17.1.4 [merge]
        committer: Baltzar Stevensson <baltzar.stevensson@mmk.su.se>
        branch nick: trajman
        timestamp: Wed 2011-10-05 12:27:12 +0200
        message:
          merged with main, enabled multiple trajectories
        ------------------------------------------------------------
        revno: 17.1.3
        committer: Baltzar Stevensson <baltzar.stevensson@mmk.su.se>
        branch nick: trajman
        timestamp: Wed 2011-10-05 11:04:18 +0200
        message:
          enabled full format of HISTORY, inithist
        ------------------------------------------------------------
        revno: 17.1.2 [merge]
        committer: Baltzar Stevensson <baltzar.stevensson@mmk.su.se>
        branch nick: trajman
        timestamp: Tue 2011-10-04 16:19:00 +0200
        message:
          added dlpoly3hist functionality
        ------------------------------------------------------------
        revno: 17.1.1
        committer: Baltzar Stevensson <baltzar.stevensson@mmk.su.se>
        branch nick: trajman
        timestamp: Mon 2011-10-03 13:24:14 +0200
        message:
          enable HISTORY, CONFIG
    ------------------------------------------------------------
    revno: 18.1.2 [merge]
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: trajsplit
    timestamp: Wed 2011-10-05 13:17:22 +0200
    message:
      Merged with dev.
------------------------------------------------------------
revno: 21
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Wed 2011-10-05 13:58:58 +0200
message:
  Added file changelog.txt to store the commit log.
------------------------------------------------------------
revno: 20
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Wed 2011-10-05 10:43:49 +0200
message:
  Added file type error test. Not allowed to mix.
------------------------------------------------------------
revno: 19 [merge]
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Tue 2011-10-04 23:04:32 +0200
message:
  Merged with the trajsplit branch.
    ------------------------------------------------------------
    revno: 18.1.1
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: trajsplit
    timestamp: Tue 2011-10-04 23:02:48 +0200
    message:
      First attempt to read trajectory files in a sequential way. Seems to work.
------------------------------------------------------------
revno: 18 [merge]
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Mon 2011-10-03 11:29:13 +0200
message:
  Turned off some debugging flags
    ------------------------------------------------------------
    revno: 7.1.51
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid
    timestamp: Fri 2011-09-30 15:55:58 +0200
    message:
      changed some debugging flags in the molfile_plugins to minimize output
------------------------------------------------------------
revno: 17
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Fri 2011-09-30 15:56:37 +0200
message:
  Changed README
------------------------------------------------------------
revno: 16 [merge]
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Fri 2011-09-30 15:35:48 +0200
message:
  Merged the area-per-lipid branch.
    ------------------------------------------------------------
    revno: 7.1.50
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid
    timestamp: Fri 2011-09-30 15:33:35 +0200
    message:
      Since the license seems to allow it, I added the vmd molfile_plugin to the source tree.
------------------------------------------------------------
revno: 15 [merge]
committer: Jon Kapla <jon.kapla@mmk.su.se>
branch nick: trajman
timestamp: Fri 2011-09-30 11:30:38 +0200
message:
  Merged branches
    ------------------------------------------------------------
    revno: 7.1.49
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid
    timestamp: Fri 2011-09-30 11:27:17 +0200
    message:
      Added some documentation. Not the best but at least some.
    ------------------------------------------------------------
    revno: 7.1.48
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid
    timestamp: Fri 2011-09-30 10:54:35 +0200
    message:
      Made trajman handle all trajectory formats that VMD can handle. Still only .gro as init.
    ------------------------------------------------------------
    revno: 7.1.47
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid
    timestamp: Fri 2011-09-09 10:48:28 +0200
    message:
      Back folding of molecules part of the centerofmembrane now seems to work. The error was due to different units in .gro init and .trr traj.
    ------------------------------------------------------------
    revno: 7.1.46
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid
    timestamp: Thu 2011-09-08 13:44:50 +0200
    message:
      ugly hacking
    ------------------------------------------------------------
    revno: 7.1.45
    committer: Jon Kapla <jon.kapla@mmk.su.se>
    branch nick: area-per-lipid

trajman.f90
!---LICENSE-------------------------------------------------------!{{{
! This file is part of
!
!  Trajman: A MD Trajectory Analysis Tool
!
!  Copyright (C) 2010, Jon Kapla, jon.kapla@mmk.su.se
!  Department of Material and Environmental Chemistry,
!  Stockholm University
!
! Trajman is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.
!
! Trajman is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License
! along with Trajman.  If not, see <http://www.gnu.org/licenses/>.
!-----------------------------------------------------------------!}}}
program trajman
    use util
    use kinds
    use input
    use readtraj
    use trajop
    use statistics
    implicit none
    character(kind=1,len=1),allocatable :: inputrad(:),words(:,:)
    character(kind=1,len=255) :: sysstr,pid,wf,filename
    integer(kind=ik) :: i,j,k,l,m,n,ios,p,frame
    integer(kind=4) :: runit
    integer(kind=8) :: trs,frs
    integer(kind=4) :: natm,statf
    real(kind=4),allocatable :: coorv(:),bx(:) !MOLFILEPLUGIN FROM VMD
    type(instruct),allocatable :: troptype(:)
    integer(kind=ik) :: v(1:4),vs(1:4)
    call RNSTRT(310952_ikr) !init. random seed
    skipframes=0;maxframes=0
    call arguments(runit)
    call constants
    call globals
    i=0
    allocate(inputrad(1))
    do !Input processing 
        call readline(inputrad,ios,runit)
        if(ios==endf)exit
        i=i+1
        call reallocinstruct(troptype,i)
        call procinp(inputrad,troptype(i))
    end do
    troptype(:)%nmolop=0
    do i=1,size(troptype)
        select case(troptype(i)%findex)
        case(0,7,9,10,15,16,17,18,19,20)
        case(13,22) !RDF
            j=moltypeofuatom(troptype(i)%atoms(1))
            k=moltypeofuatom(troptype(i)%atoms(2))
            select case(troptype(i)%set%leaflet)
            case(0)!Both
                l=molt(j)%nmol*molt(k)%nmol
            case(1)!Lower
                l=size(molt(j)%lower)*size(molt(k)%lower)
            case(2)!Upper
                l=size(molt(j)%upper)*size(molt(k)%upper)
            end select
            !if(j==k)l=nint(sqrt(real(l,rk))*(sqrt(real(l,rk))-1))/2
            if(j==k)l=nint(sqrt(real(l,rk))*(sqrt(real(l,rk))-1)/2)
            if(troptype(i)%findex==13)allocate(troptype(i)%rdf_pairs(l))
            if(troptype(i)%findex==22)troptype(i)%nmolop=l
           ! write(*,*)troptype(i)%nmolop,troptype(i)%findex,j,k,molt(1)%nmol
           ! stop

                
        case default
            j=moltypeofuatom(troptype(i)%atoms(1))
            select case(troptype(i)%set%leaflet)
                case(0)
                    call reallocate(troptype(i)%molind,molt(j)%nmol)
                    do k=1,molt(j)%nmol
                        troptype(i)%molind(k)=k
                    end do
                case(1)
                    call reallocate(troptype(i)%molind,size(molt(j)%lower))
                    troptype(i)%molind=molt(j)%lower
                case(2)
                    call reallocate(troptype(i)%molind,size(molt(j)%upper))
                    troptype(i)%molind=molt(j)%upper
            end select
            troptype(i)%nmolop=size(troptype(i)%molind)
        end select
    end do
    ! Allocate data matrix to a size based on input
    if(maxframes==0)then
        select case(trajtype)
        case('gro')
            !Sum the number of frames of all trajfiles
            j=0
            do i=1,size(trajfile)
            p=getpid() ! Returns pid of the running instance of trajman
            write(pid,*)p
            !Make a system call to stat to get filesize of trajectory file
            write(sysstr,*)'stat -c%s '//stringconv(trajfile(i)%filename)//' >',trim(adjustl(pid)),'trsize'
            ios=system(trim(sysstr))
            !Make a system call to extract one frame from trajectory file (based on
            !rowsperframe)
            write(sysstr,*)'tail ',-rowsperframe,' ',stringconv(trajfile(i)%filename),' >',trim(adjustl(pid)),'oneframe'
            ios=system(trim(sysstr))
            ! Make a system call to get the size of one frame
            sysstr='stat -c%s '//trim(adjustl(pid))//'oneframe >>'//trim(adjustl(pid))//'trsize'
            ios=system(trim(sysstr))
            !Get sizes and calculate number of frames
            open(42,file=trim(adjustl(pid))//'trsize')
            read(42,*)trs
            read(42,*)frs
            j=((trs-mod(trs,frs))/frs+nint(real(mod(trs,frs),8)/real(frs,8)))
            maxframes=maxframes+j
            if(trs==0)maxframes=0
            close(42)
            ! Remove temporary files
            ios=system('rm '//trim(adjustl(pid))//'trsize '//trim(adjustl(pid))//'oneframe')
            if(.NOT. common_setflags%silent)write(0,"(A,5x,A15,I6)",advance="no")&
           char(13),"Counting frames: ",j!,char(13)     
            write(*,*)stringconv(trajfile(i)%filename)
            end do
            
        case('trr','dcd','pdb')!VMD MOLFILEPLUGIN
            statf=1
            allocate(coorv(atot*3),bx(6))
            do i=1,size(tunit)
                j=0
                do 
                    call f77_molfile_read_next(tunit(i),int(atot,4),coorv,bx,statf)
                    if(.NOT. common_setflags%silent)write(0,"(A,5x,A15,I6)",advance="no")&
                    char(13),"Counting frames: ",j!,char(13)     
                    if(statf==0)exit
                    maxframes=maxframes+1
                    j=j+1
                end do
                if(.NOT. common_setflags%silent)write(*,*)stringconv(trajfile(i)%filename)
                call f77_molfile_close_read(tunit(i),statf)
                call f77_molfile_open_read(tunit(i),natm,stringconv(trajfile(i)%filename),trajtype)
            end do
            deallocate(coorv,bx)
         case(dlpoly3histfile(1:min(len(dlpoly3histfile),len(trajtype))))
            statf=1
            maxframes=0
            do i=1,size(tunit)
               call inithist(trajfile(i)%filename,j)
               maxframes=maxframes+j
               if(.NOT. common_setflags%silent)then
                  write(0,"(A,5x,A15,I6)",advance="no")&
                       char(13),"Counting frames: ",j
                  write(0,*)stringconv(trajfile(i)%filename)
               end if
            end do
        case default
            statf=1
            allocate(coorv(atot*3),bx(6))
            do i=1,size(tunit)
                j=0
                do 
                    call f77_molfile_read_next(tunit(i),int(atot,4),coorv,bx,statf)
                    if(.NOT. common_setflags%silent)write(0,"(A,5x,A15,I6)",advance="no")&
                    char(13),"Counting frames: ",j!,char(13)     
                    if(statf==0)exit
                    maxframes=maxframes+1
                    j=j+1
                end do
                if(.NOT. common_setflags%silent)write(*,*)stringconv(trajfile(i)%filename)
                call f77_molfile_close_read(tunit(i),statf)
                call f77_molfile_open_read(tunit(i),natm,stringconv(trajfile(i)%filename),'auto')
            end do
            deallocate(coorv,bx)
        end select
    end if
    do i=1,size(troptype)
        if(troptype(i)%nmolop*maxframes/=0)then
            allocate(troptype(i)%datam(troptype(i)%nmolop,skipframes+1:maxframes))
            troptype(i)%datam=0
        else
            select case(troptype(i)%findex)
            case(13)
                allocate(troptype(i)%rdf_dist(troptype(i)%set%distbin))
                allocate(troptype(i)%datam(1,skipframes+1:maxframes))
                troptype(i)%rdf_dist=0
            case(15)
                allocate(troptype(i)%datam(1,skipframes+1:maxframes))
                troptype(i)%datam=0
            case(16)
                !allocate(troptype(i)%rdf_dist(troptype(i)%set%distbin))
                !troptype(i)%rdf_dist=0
                !troptype(i)%rdf_bin=1._rk/troptype(i)%set%distbin
            case(17,18)
                allocate(troptype(i)%datam(1,skipframes+1:maxframes))
                troptype(i)%datam=0
            end select
        end if
    end do
    !write(*,*)size(troptype(18)%datam,2),'DATAM';stop
    if(.NOT. common_setflags%silent)then
        write(*,*)"input processing done."
        call summary
        write(*,'(5X,A19,I6,A2,I6)')'Frames to process: '&
        ,maxframes-skipframes," /",maxframes
    end if
!###########################FRAME LOOP##################################################    
j=1;i=0
do !i=1,skipframes
    if(i==skipframes)exit
    i=i+1
    if(.NOT. common_setflags%silent)write(0,"(A,5x,A15,I6)",advance="no")&
    char(13),"Skipping frame: ",i!,char(13) 
    ios=readframe(tunit(j))
    if(ios==2)then
        j=j+1 ! Change file!
        i=i-1 ! Rewind the counter so that we get the last frame
    end if
end do
if(skipframes/=0.and..NOT.common_setflags%silent)write(*,*)&
"    Starting calculation on frame: ",trim(adjustl(intstr(skipframes+1)))
frame=skipframes
    do k=j,size(tunit) !Loop over trajectory files
    do while (readframe(tunit(k))/=2)! Trajectory processing, frameloop
   ! write(*,*)tunit(k),'TUNIT'
        frame=frame+1
        if(modulo(frame-skipframes,max((maxframes-skipframes)/100,1))==0.OR.&
        frame==skipframes+1)then
            if(.NOT. common_setflags%silent)then
                write(0,'(A,5X,A10,I3,A2,I6)',advance='no')&
                char(13),'Progress: ',nint(real(100*(frame-skipframes),rk)/real(maxframes-skipframes,rk)),'% ',frame!,char(13)
            end if
        end if
        if(global_setflags%whole)call whole
        if(global_setflags%folding)call foldmol
        meanbox(1:3)=meanbox(1:3)+box
        meanbox(4)=meanbox(4)+1/(box(1)*box(2)*box(3))
        call procop(troptype,frame) ! CALCULATIONS: Perform instructions on frame
        if(frame==skipframes+1)then !Write atomnames and coordinates for the first molecules
                        !to a .xyz file
            open(37,file='atoms.xyz')
            write(37,*)size(atomnames)
            write(37,*)
            do i=1,size(atomnames)
                select case(trajtype)
                case('gro')
                    write(37,*)atomnames(i),10*coor(1:3,cind(atomindex(atomnames(i),atomnames,size(atomnames)),1_ik))
                case('trr','dcd','pdb')
                    write(37,*)atomnames(i),coor(1:3,cind(atomindex(atomnames(i),atomnames,size(atomnames)),1_ik))
                case default
                    write(37,*)atomnames(i),coor(1:3,cind(atomindex(atomnames(i),atomnames,size(atomnames)),1_ik))
                end select
            end do
            close(37)
        end if

        do m=1,global_setflags%wftot !BEGIN WRITEFRAME: Write specified frame to file
            if(allocated(global_setflags%writeframe) .AND. frame==global_setflags%writeframe(m)%framenumber)then
                write(wf,*)global_setflags%writeframe(m)%framenumber
                
                select case(trim(global_setflags%writeframe(m)%outformat))
                case('gro')
                    filename='frame'//trim(adjustl(wf))//'.gro'
                    call wf_gro(trim(filename),wf,37)
                case('xyz') ! Defaults to writing a .xyz file with long atomnames,
                     ! viewable by gOpenmol.
                    filename='frame'//trim(adjustl(wf))//'.xyz'
                    call wf_xyz(trim(filename),37)
                end select
            end if
        end do !END WRITEFRAME       
        if (frame==maxframes)exit
        end do ! Frameloop
        if(frame==maxframes)exit
        end do ! Trajectory file loop
!#######################################################################################
    do i=1,size(tunit)
        call closetraj(tunit(i))
    end do
    if(.NOT. common_setflags%silent)then
        write(*,*)
        write(*,'(5x,A)')'Postprocessing...'
    end if
    call postproc(troptype) ! POSTPROCESSING
    if(.NOT. common_setflags%silent)write(*,'(5X,A)')'Done!'
end program trajman
